File: backend/services/docker_metrics.py
import docker
from datetime import datetime
from typing import List, Dict, Optional
import os

class DockerMetricsService:
    def __init__(self):
        try:
            # Try to connect using Unix socket directly
            self.client = docker.DockerClient(base_url='unix://var/run/docker.sock')
            self.client.ping()
            print("✅ Connected to Docker daemon via Unix socket")
        except Exception as e:
            print(f"❌ Failed to connect to Docker: {e}")
            print("Trying alternative connection method...")
            try:
                # Alternative: Try using environment variables
                self.client = docker.from_env()
                self.client.ping()
                print("✅ Connected to Docker daemon via environment")
            except Exception as e2:
                print(f"❌ Both connection methods failed: {e2}")
                self.client = None

    def get_student_containers(self, user_id: Optional[str] = None) -> List[Dict]:
        """Get all student-deployed containers"""
        if not self.client:
            return []
        
        try:
            filters = {'label': 'deployed_by=student'}
            if user_id:
                filters['label'] = [f'deployed_by=student', f'user_id={user_id}']
            
            containers = self.client.containers.list(all=True, filters=filters)
            
            result = []
            for container in containers:
                result.append({
                    'id': container.short_id,
                    'name': container.name,
                    'image': container.image.tags[0] if container.image.tags else 'unknown',
                    'status': container.status,
                    'created': container.attrs['Created'],
                    'labels': container.labels,
                    'ports': container.ports
                })
            
            return result
        except Exception as e:
            print(f"Error getting containers: {e}")
            return []

    def get_container_metrics(self, container_id: str) -> Dict:
        """Get real-time metrics for a specific container"""
        if not self.client:
            return {'error': 'Docker not connected'}
        
        try:
            container = self.client.containers.get(container_id)
            
            if container.status != 'running':
                return {
                    'containerId': container_id,
                    'running': False,
                    'cpu': 0,
                    'memory': 0,
                    'error': 'Container is not running'
                }
            
            # Get stats (one-shot)
            stats = container.stats(stream=False)
            
            # Calculate CPU percentage
            cpu_delta = stats['cpu_stats']['cpu_usage']['total_usage'] - \
                       stats['precpu_stats']['cpu_usage']['total_usage']
            system_delta = stats['cpu_stats']['system_cpu_usage'] - \
                          stats['precpu_stats']['system_cpu_usage']
            cpu_count = stats['cpu_stats']['online_cpus']
            
            cpu_percent = 0
            if system_delta > 0 and cpu_delta > 0:
                cpu_percent = (cpu_delta / system_delta) * cpu_count * 100
            
            # Calculate Memory percentage
            memory_usage = stats['memory_stats'].get('usage', 0)
            memory_limit = stats['memory_stats'].get('limit', 1)
            memory_percent = (memory_usage / memory_limit) * 100
            
            # Network stats
            networks = stats.get('networks', {})
            network_rx = sum(net.get('rx_bytes', 0) for net in networks.values())
            network_tx = sum(net.get('tx_bytes', 0) for net in networks.values())
            
            return {
                'containerId': container_id[:12],
                'running': True,
                'cpu': round(min(cpu_percent, 100), 2),
                'memory': round(min(memory_percent, 100), 2),
                'memoryUsage': self.format_bytes(memory_usage),
                'memoryLimit': self.format_bytes(memory_limit),
                'networkRx': self.format_bytes(network_rx),
                'networkTx': self.format_bytes(network_tx),
                'timestamp': datetime.now().isoformat()
            }
        except Exception as e:
            print(f"Error getting metrics for {container_id}: {e}")
            return {
                'containerId': container_id,
                'running': False,
                'cpu': 0,
                'memory': 0,
                'error': str(e)
            }

    def get_aggregated_metrics(self, user_id: Optional[str] = None) -> Dict:
        """Get aggregated metrics for all student containers"""
        containers = self.get_student_containers(user_id)
        
        if not containers:
            return {
                'avgCpu': 0,
                'avgMemory': 0,
                'totalContainers': 0,
                'runningContainers': 0,
                'stoppedContainers': 0,
                'containers': []
            }
        
        running = [c for c in containers if c['status'] == 'running']
        metrics_list = [self.get_container_metrics(c['id']) for c in running]
        valid_metrics = [m for m in metrics_list if m.get('running')]
        
        avg_cpu = sum(m['cpu'] for m in valid_metrics) / len(valid_metrics) if valid_metrics else 0
        avg_memory = sum(m['memory'] for m in valid_metrics) / len(valid_metrics) if valid_metrics else 0
        
        containers_with_metrics = []
        for container in containers:
            metric = next((m for m in metrics_list if m['containerId'] == container['id']), None)
            containers_with_metrics.append({
                **container,
                'metrics': metric or {'cpu': 0, 'memory': 0, 'running': False}
            })
        
        return {
            'avgCpu': round(avg_cpu, 2),
            'avgMemory': round(avg_memory, 2),
            'totalContainers': len(containers),
            'runningContainers': len(running),
            'stoppedContainers': len(containers) - len(running),
            'containers': containers_with_metrics
        }

    def get_container_logs(self, container_id: str, tail: int = 100) -> str:
        """Get container logs"""
        if not self.client:
            return "Docker not connected"
        
        try:
            container = self.client.containers.get(container_id)
            logs = container.logs(tail=tail, timestamps=True)
            return logs.decode('utf-8')
        except Exception as e:
            return f"Error: {str(e)}"

    @staticmethod
    def format_bytes(bytes_val: int) -> str:
        """Format bytes to human readable"""
        if bytes_val == 0:
            return "0 B"
        
        sizes = ['B', 'KB', 'MB', 'GB', 'TB']
        k = 1024
        i = 0
        while bytes_val >= k and i < len(sizes) - 1:
            bytes_val /= k
            i += 1
        
        return f"{round(bytes_val, 2)} {sizes[i]}"

    def check_docker_status(self) -> Dict:
        """Check Docker daemon status"""
        if not self.client:
            return {'running': False, 'error': 'Not connected'}
        
        try:
            info = self.client.info()
            return {
                'running': True,
                'containers': info['Containers'],
                'containersRunning': info['ContainersRunning'],
                'containersPaused': info['ContainersPaused'],
                'containersStopped': info['ContainersStopped'],
                'images': info['Images'],
                'serverVersion': info['ServerVersion']
            }
        except Exception as e:
            return {'running': False, 'error': str(e)}


# Create singleton instance
docker_metrics_service = DockerMetricsService()

File: backend/services/dockerMetrics.js
const Docker = require('dockerode');
const docker = new Docker({ socketPath: '/var/run/docker.sock' });

class DockerMetricsService {
  constructor() {
    this.metricsCache = new Map();
  }

  async getStudentContainers(userId = null) {
    try {
      const filters = {
        label: ['deployed_by=student']
      };
      
      if (userId) {
        filters.label.push(`user_id=${userId}`);
      }

      const containers = await docker.listContainers({
        all: true,
        filters: filters
      });

      return containers.map(container => ({
        id: container.Id.substring(0, 12),
        fullId: container.Id,
        name: container.Names[0].replace('/', ''),
        image: container.Image,
        status: container.State,
        state: container.Status,
        created: new Date(container.Created * 1000).toISOString(),
        ports: this.formatPorts(container.Ports),
        labels: container.Labels
      }));
    } catch (error) {
      console.error('Error getting student containers:', error);
      return [];
    }
  }

  async getContainerMetrics(containerId) {
    try {
      const container = docker.getContainer(containerId);
      
      const inspection = await container.inspect();
      if (!inspection.State.Running) {
        return {
          containerId,
          running: false,
          cpu: 0,
          memory: 0,
          error: 'Container is not running'
        };
      }

      const stats = await container.stats({ stream: false });

      const cpuDelta = stats.cpu_stats.cpu_usage.total_usage - 
                       (stats.precpu_stats.cpu_usage?.total_usage || 0);
      const systemDelta = stats.cpu_stats.system_cpu_usage - 
                          (stats.precpu_stats.system_cpu_usage || 0);
      const cpuCount = stats.cpu_stats.online_cpus || 1;
      
      let cpuPercent = 0;
      if (systemDelta > 0 && cpuDelta > 0) {
        cpuPercent = (cpuDelta / systemDelta) * cpuCount * 100;
      }

      const memoryUsage = stats.memory_stats.usage || 0;
      const memoryLimit = stats.memory_stats.limit || 1;
      const memoryPercent = (memoryUsage / memoryLimit) * 100;

      const networks = stats.networks || {};
      let networkRx = 0;
      let networkTx = 0;
      Object.values(networks).forEach(net => {
        networkRx += net.rx_bytes || 0;
        networkTx += net.tx_bytes || 0;
      });

      const metrics = {
        containerId: containerId.substring(0, 12),
        running: true,
        cpu: Math.min(Math.round(cpuPercent * 100) / 100, 100),
        memory: Math.min(Math.round(memoryPercent * 100) / 100, 100),
        memoryUsage: this.formatBytes(memoryUsage),
        memoryLimit: this.formatBytes(memoryLimit),
        memoryUsageBytes: memoryUsage,
        memoryLimitBytes: memoryLimit,
        networkRx: this.formatBytes(networkRx),
        networkTx: this.formatBytes(networkTx),
        networkRxBytes: networkRx,
        networkTxBytes: networkTx,
        timestamp: new Date().toISOString()
      };

      this.metricsCache.set(containerId, metrics);
      
      return metrics;
    } catch (error) {
      console.error(`Error getting metrics for container ${containerId}:`, error.message);
      return {
        containerId,
        running: false,
        cpu: 0,
        memory: 0,
        error: error.message
      };
    }
  }

  async getAggregatedMetrics(userId = null) {
    try {
      const containers = await this.getStudentContainers(userId);
      
      if (containers.length === 0) {
        return {
          avgCpu: 0,
          avgMemory: 0,
          totalContainers: 0,
          runningContainers: 0,
          stoppedContainers: 0,
          containers: []
        };
      }

      const runningContainers = containers.filter(c => c.status === 'running');
      const metricsPromises = runningContainers.map(c => this.getContainerMetrics(c.fullId));
      const metrics = await Promise.all(metricsPromises);
      const validMetrics = metrics.filter(m => m.running && !m.error);

      const avgCpu = validMetrics.length > 0 
        ? validMetrics.reduce((sum, m) => sum + m.cpu, 0) / validMetrics.length 
        : 0;
      const avgMemory = validMetrics.length > 0 
        ? validMetrics.reduce((sum, m) => sum + m.memory, 0) / validMetrics.length 
        : 0;

      const containersWithMetrics = containers.map(container => {
        const metric = metrics.find(m => m.containerId === container.id) || {
          cpu: 0,
          memory: 0,
          running: container.status === 'running'
        };
        return {
          ...container,
          metrics: metric
        };
      });

      return {
        avgCpu: Math.round(avgCpu * 100) / 100,
        avgMemory: Math.round(avgMemory * 100) / 100,
        totalContainers: containers.length,
        runningContainers: runningContainers.length,
        stoppedContainers: containers.length - runningContainers.length,
        containers: containersWithMetrics
      };
    } catch (error) {
      console.error('Error getting aggregated metrics:', error);
      return {
        avgCpu: 0,
        avgMemory: 0,
        totalContainers: 0,
        runningContainers: 0,
        stoppedContainers: 0,
        containers: [],
        error: error.message
      };
    }
  }

  async getContainerLogs(containerId, tail = 100) {
    try {
      const container = docker.getContainer(containerId);
      const logs = await container.logs({
        stdout: true,
        stderr: true,
        tail: tail,
        timestamps: true
      });
      
      return logs.toString('utf-8');
    } catch (error) {
      console.error(`Error getting logs for container ${containerId}:`, error);
      return `Error: ${error.message}`;
    }
  }

  formatBytes(bytes) {
    if (bytes === 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB', 'TB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
  }

  formatPorts(ports) {
    if (!ports || ports.length === 0) return [];
    return ports.map(port => ({
      private: port.PrivatePort,
      public: port.PublicPort || null,
      type: port.Type
    }));
  }

  async checkDockerStatus() {
    try {
      await docker.ping();
      const info = await docker.info();
      return {
        running: true,
        containers: info.Containers,
        containersRunning: info.ContainersRunning,
        containersPaused: info.ContainersPaused,
        containersStopped: info.ContainersStopped,
        images: info.Images,
        serverVersion: info.ServerVersion
      };
    } catch (error) {
      return {
        running: false,
        error: error.message
      };
    }
  }
}

module.exports = new DockerMetricsService();
